### PR TITLE
feat(distro-wildfly): use jboss-logging & JSON logging examples

### LINKED ISSUE #1112: Add JSON logging config to Wildfly & EAP distros

### PR DESCRIPTION
**Pull request by msavy**
Closes #1112

### CODE REVIEW

File: distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml [hunk 1]
```diff
@@ -132,6 +141,9 @@
             <logger category="sun.rmi">
                 <level name="WARN"/>
             </logger>
+            <logger category="io.apiman">
```
[the remark this answers is not in the document]
Comment (msavy) [reply distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml:144]: Yes, Apiman wasn't directly using `jboss-logging` before, so now you can configure a different log level for io.apiman (or whatever you like).

File: distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml [hunk 2]
```diff
@@ -41,6 +41,11 @@
         <extension module="org.keycloak.keycloak-server-subsystem"/>
         <extension module="org.keycloak.keycloak-adapter-subsystem"/>
     </extensions>
+    <system-properties>
+        <!-- This property tells Apiman log via jboss logging. Apiman-specific logging config can
+        be set via the logging subsystem in this standalone.xml file. -->
+        <property name="apiman.logger-delegate" value="jboss-logging"/>
```
[the remark this answers is not in the document]
Comment (msavy) [reply distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml:47]: You can set already system properties via an environment variable simply by adding them to the JAVA_OPTS env var

As an example, you could do....

`JAVA_OPTS="-Dapiman.logger-delegate=jboss-logging"`

I don't particularly expect anyone will want to change this one 🤔. Could be interesting for other cases though.

File: manager/test/api/src/main/java/io/apiman/manager/test/junit/ManagerRestTester.java [hunk 3]
```diff
@@ -54,7 +56,13 @@
 @SuppressWarnings("nls")
 public class ManagerRestTester extends ParentRunner<TestInfo> {
 
-    private static Logger logger = LoggerFactory.getLogger(TestPlanRunner.class);
+    private static final Logger LOGGER;
+
+    static {
```
Comment (msavy) [inline manager/test/api/src/main/java/io/apiman/manager/test/junit/ManagerRestTester.java:61]: Can just leave this as default

File: common/logging/core/src/main/java/io/apiman/common/logging/MultiLogger.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
+package io.apiman.common.logging;
```
Comment (msavy) [inline common/logging/core/src/main/java/io/apiman/common/logging/MultiLogger.java:1]: headers

File: common/logging/core/src/main/java/io/apiman/common/logging/DoubleLogger.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
+package io.apiman.common.logging;
```
Comment (msavy) [inline common/logging/core/src/main/java/io/apiman/common/logging/DoubleLogger.java:1]: headers

**Comment by msavy** [comment]
This PR morphed into a bit more than originally intended (by necessity 😂). Will ensure the commits reflect that in some way.

File: distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (volkflo) [reply distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml:144]: Cool, like that :)

File: distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (volkflo) [reply distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml:47]: I am not a big fan of this JAVA_OPTS thing, not nice for docker-compose and kubernetes, because you have to maintain one big string with all you options. But for the moment I think it is okay to leave it as is and change it when there is need for it :)

File: common/logging/core/src/main/java/io/apiman/common/logging/impl/DoubleLogger.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+ * Combine two loggers together!
+ *
+ * Can be useful for testing and if logging output should go to somewhere in addition to the logging
+ * subsystem (e.g. response body).
+ *
+ * @author Marc Savy {@literal <marc@blackparrotlabs.io>}
+ */
+public class DoubleLogger implements IApimanLogger {
+
+    private final IApimanLogger loggerA;
```
Comment (volkflo) [inline common/logging/core/src/main/java/io/apiman/common/logging/impl/DoubleLogger.java:31]: Are these two loggers for logging to stdout (console) and one to log into a file?

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/ExportImportManager.java [hunk 7]
```diff
@@ -36,10 +36,8 @@
     
     @Inject
     private ExportImportConfigParser config;
-    @Inject @ApimanLogger(IImportReader.class)
-    private IApimanLogger importLogger;
-    @Inject @ApimanLogger(IExportWriter.class)
-    private IApimanLogger exportLogger;
+    private IApimanLogger importLogger = ApimanLoggerFactory.getLogger(IImportReader.class);
```
Comment (volkflo) [inline manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/ExportImportManager.java:39]: Any reason to not make this also `static final`?

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageExporter.java [hunk 8]
```diff
@@ -39,19 +40,18 @@
...
 import java.util.Date;
 import java.util.Iterator;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
 
 @ApplicationScoped
 public class StorageExporter {
-    @Inject @ApimanLogger(StorageExporter.class)
-    private IApimanLogger logger;
+    private final IApimanLogger logger = ApimanLoggerFactory.getLogger(StorageExporter.class);
```
Comment (volkflo) [inline manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageExporter.java:53]: Any reason to not make this also static final?

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageImportDispatcher.java [hunk 9]
```diff
@@ -68,8 +79,8 @@
 
     @Inject
     private IStorage storage;
-    @Inject @ApimanLogger(StorageImportDispatcher.class)
-    private IApimanLogger logger;
+    private IApimanLogger logger = ApimanLoggerFactory.getLogger(StorageImportDispatcher.class);
```
Comment (volkflo) [inline manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageImportDispatcher.java:82]: Any reason to not make this also static final?

File: manager/api/migrator/src/main/java/io/apiman/manager/api/migrator/DataMigrator.java [hunk 10]
```diff
@@ -39,10 +40,10 @@
 @Dependent
 public class DataMigrator implements IReaderHandler {
 
+    private IApimanLogger logger = ApimanLoggerFactory.getLogger(DataMigrator.class);
```
Comment (volkflo) [inline manager/api/migrator/src/main/java/io/apiman/manager/api/migrator/DataMigrator.java:43]: Any reason to not make this also static final?

**Review by volkflo — APPROVED** [review]
Maybe a global question/note:
Is there a reason why not all loggers are `static final`?

The import changes make it quite hard to follow the changes.
Would be nice to move the import refactoring into a new commit ?

File: common/logging/core/src/main/java/io/apiman/common/logging/impl/DoubleLogger.java (hunk 6 above)
Comment (msavy) [reply common/logging/core/src/main/java/io/apiman/common/logging/impl/DoubleLogger.java:31]: Logging out to both console and to the response body of the request (see: https://github.com/msavy/apiman/blob/add-jboss-logging-to-wf-eap-distros/manager/api/rest-impl/src/main/java/io/apiman/manager/api/rest/impl/SystemResourceImpl.java#L177-L244)

This provides a backwards compatible mechanism of achieving this (so we get both to the logging framework and to the alternative source).

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/ExportImportManager.java (hunk 7 above)
Comment (msavy) [reply manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/ExportImportManager.java:39]: Refer to the previous comment. 

I did this initially and had to back it out. It turns out that the implementation of ImportExport uses the logger to collect some logging information that is put into the HTTP response body when you do an import/export via https://github.com/msavy/apiman/blob/add-jboss-logging-to-wf-eap-distros/manager/api/rest-impl/src/main/java/io/apiman/manager/api/rest/impl/SystemResourceImpl.java#L177-L244

By changing this to `private final static` this trick doesn't work properly any more. 

It's an unusual use of a logger, but I want to just maintain compatibility for now.

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageExporter.java (hunk 8 above)
Comment (msavy) [reply manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageExporter.java:53]: See previous

File: manager/api/migrator/src/main/java/io/apiman/manager/api/migrator/DataMigrator.java (hunk 10 above)
Comment (msavy) [reply manager/api/migrator/src/main/java/io/apiman/manager/api/migrator/DataMigrator.java:43]: See previous

File: manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageImportDispatcher.java (hunk 9 above)
Comment (msavy) [reply manager/api/export-import/src/main/java/io/apiman/manager/api/exportimport/manager/StorageImportDispatcher.java:82]: See previous

File: distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (msavy) [reply distro/wildfly/src/main/resources/overlay/standalone/configuration/standalone-apiman.xml:47]: I'll speak to some of my friends who dockerised a bunch of projects and ask what their solution to this was. 

I can think of a few solutions off the top of my head, including having a script that just translates environment variables into system properties or similar.

Really want to avoid having to support both of these at the same time.

They also have subtly different behaviours: system properties are mutable in Java, whereas the environment properties are immutable.
